### PR TITLE
Payload [WIP]

### LINKED ISSUE #93: Payload game mode
git@github.com:StratusNetwork/projectares.git has implemented a game mode called `Payload`.

Players have to push a Minecart (known as a “Payload”) to the end of their track. There are two setups for this gamemode: Attackers vs. Defenders, and Team vs. Team. Both versions are supported with the following modules.

GitHub source: https://github.com/StratusNetwork/projectares/tree/master/PGM/src/main/java/tc/oc/pgm/payload
XML documentation: http://web.archive.org/web/20181028210707/https://docs.stratus.network/modules/gamemode_payload

### PR DESCRIPTION
**Pull request by KingOfSquares**
Welcome. Should close #93 

I marked this as WIP since I need to test some scenarios to be sure I don't release a broken gamemode, ~~but I don't have much more to add code-wise. One problem I have is that the XML can become quite packed with attributes for a single payload if you have a lot of custom values.~~ Releasing this to _hopefully_  get some feedback since this is the first time I've worked on something so "big".

I'll add some video later

**Some XML examples:**

<details>
<summary>A very simple XML can look like this</summary>

```xml
<payloads>
    <payload name="Payload" id="payload">
        <goals>
            <goal team="blue-team">59.5,22.0,32.5</goal>
            <goal team="red-team">-58.5,22.0,32.5</goal>
        </goals>
        <starting-location>0.5,22.0,0.5</starting-location>
    </payload>
</payloads>
```
</details>
<details>
<summary>A more custom XML could looks a little like this</summary>

```xml
<payloads>
    <payload name="Payload" id="payload">
        <goals>
            <goal team="blue-team">59.5,22.0,32.5</goal>
            <goal team="red-team">-58.5,22.0,32.5</goal>
        </goals>
        <properties radius="5" height="8"/>
        <speeds neutral-speed="0.5">
            <speed team="red-team">1.1</speed>
            <speed team="blue-team">1.1</speed>
        </speeds>
        <starting-location>0.5,22.0,0.5</starting-location>
        <checkpoints>
            <checkpoint id="blue-1" permanent="true">28.5,22.00,-13.5</checkpoint>
            <checkpoint id="blue-2">52.5,22.0,16.5</checkpoint>
            <checkpoint id="red-1" permanent="true">-27.5,22.00,-13.5</checkpoint>
            <checkpoint id="red-2">-51.5,22.0,16.5</checkpoint>
        </checkpoints>
    </payload>
</payloads>
```

</details>

**Future**

Features **to be** implemented

- [ ] Progressive faster speed based on the amount of players on the payload(2 pushing should be faster than 1)

<details>
<summary>Stuff I need to test with multiple people to ensure I didn't break anything</summary>

- [ ] ControlPoint maps

- [ ] Payload with multiple people

- [ ] Payload on another map than Skyline

- [ ] Payload with a single owner

- [ ] Payload with two owners, but where only the primary team can complete the goal(secondary team should win if time runs out)

</details>

<details>
<summary>Things to do *after* this has been merged</summary>

- [ ] use `RegionPlayerTracker` for spawners

- [ ] change /match so shared goals don't display multiple times

</details>

### CODE REVIEW

**Comment by BGMP** [comment]
pair programming

**Comment by KingOfSquares** [comment]
@Pablete1234 I fixed the issue you mentioned with the XML, it also gave me an opportunity to rework how the pathfinding worked a little. I like the new XML syntax much better now :)

File: core/src/main/java/tc/oc/pgm/payload/Path.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
+package tc.oc.pgm.payload;
+
+import org.bukkit.Location;
+
+public class Path {
```
Comment (Pablete1234) [inline core/src/main/java/tc/oc/pgm/payload/Path.java:5]: This is just a re-implementation of a LinkedList, use a list iterator for this instead

**Comment by Pugzy** [comment]
What is the reason for the `properties` element rather than using attributes on the main `payload` element? 
Not sure I've seen this done in any other module so looks a bit out of place.

File: core/src/main/java/tc/oc/pgm/payload/Path.java (hunk 1 above)
Comment (KingOfSquares) [reply core/src/main/java/tc/oc/pgm/payload/Path.java:5]: 🤦 you're right lol

**Comment by KingOfSquares** [comment]
> What is the reason for the `properties` element rather than using attributes on the main `payload` element?
> Not sure I've seen this done in any other module so looks a bit out of place.

This was originally to reduce element attributes, now that I have removed some other attributes I should add this back as attributes on the main payload element

**Pull request closed without merging**